Compilation fails with GCC 7.3

Hi @Pro,
I have some issues with GCC-7.3 cross-compiling the sources under Yocto.
Its fails w/ the following error messages:
```
/XXX/open62541/git-r0/recipe-sysroot-native=   -march=armv5e -marm  --sysroot=/XXX/open62541/git-r0/recipe-sysroot -O2 -DNDEBUG   -std=c99 -pipe -Wall -Wextra -Werror -Wpedantic -Wno-overlength-strings -Wno-unused-parameter -Wmissing-prototypes -Wstrict-prototypes -Wredundant-decls -Wformat -Wformat-security -Wformat-nonliteral -Wuninitialized -Winit-self -Wcast-qual -Wstrict-overflow -Wnested-externs -Wmultichar -Wundef -Wc++-compat -Wshadow -Wconversion -fvisibility=hidden -fPIC -ffunction-sections -fdata-sections -fno-stack-protector -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-math-errno -fmerge-all-constants -fno-ident -MD -MT CMakeFiles/open62541-object.dir/deps/mdnsd/libmdnsd/1035.c.o -MF CMakeFiles/open62541-object.dir/deps/mdnsd/libmdnsd/1035.c.o.d -o CMakeFiles/open62541-object.dir/deps/mdnsd/libmdnsd/1035.c.o   -c /XXX/open62541/git-r0/git/deps/mdnsd/libmdnsd/1035.c
| /XXX/open62541/git-r0/git/deps/mdnsd/libmdnsd/1035.c: In function '_rrparse':
| /XXX/open62541/git-r0/git/deps/mdnsd/libmdnsd/1035.c:270:54: error: 'snprintf' output may be truncated before the last format character [-Werror=format-truncation=]
|     msnds_snprintf(rr[i].known.a.name,15, "%d.%d.%d.%d", (*bufp)[0], (*bufp)[1], (*bufp)[2], (*bufp)[3]);
|                                                       ^
| /XXX/open62541/git-r0/git/deps/mdnsd/libmdnsd/1035.c:21:24: note: 'snprintf' output between 8 and 16 bytes into a destination of size 15
|  #define msnds_snprintf snprintf
|                         ^
| /XXX/open62541/git-r0/git/deps/mdnsd/libmdnsd/1035.c:270:4: note: in expansion of macro 'msnds_snprintf'
|     msnds_snprintf(rr[i].known.a.name,15, "%d.%d.%d.%d", (*bufp)[0], (*bufp)[1], (*bufp)[2], (*bufp)[3]);
|     ^~~~~~~~~~~~~~
| cc1: all warnings being treated as errors
```

My first analysis lead to following:

1. GCC could auto enable ``format-truncation`` (as not in explicit parameter definition list above). See https://gcc.gnu.org/gcc-7/changes.html
2. The line 270 has maybe an error, too: ``snprintf`` needs the whole buffer length, including zero-termination character. So ``size_t n`` parameter should be 16!?

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,47 +1,120 @@
-language: c
-
-compiler:
-- gcc
-- clang
-
-os:
-- linux
-- osx
+# using c for language overwrites our compilers
+language: generic
 
 dist: trusty
-sudo: required
-
-env:
-  matrix:
-    # Do both, compile with static code analysis and without
-    - ANALYZE=false
-    - ANALYZE=true
+sudo: false
 
 matrix:
-  exclude:
-  # This excludes OSX GCC builds.
+  include:
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - binutils-mingw-w64-i686
+            - build-essential
+            - gcc-multilib
+            - gcc-4.9
+            - gcc-mingw-w64-i686
+            - g++-4.9
+            - g++-multilib
+            - g++-mingw-w64
+            - libc6-dev-i386
+            - linux-libc-dev:i386
+            - mingw-w64
+      env:
+         - ANALYZE=false
+         - MINGW=true
+         - CC=gcc-4.9
+         - CXX=g++-4.9
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - gcc-7
+            - gcc-7-multilib
+            - g++-7
+            - g++-7-multilib
+      env:
+         - ANALYZE=false
+         - CC=gcc-7
+         - CXX=g++-7
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - llvm-toolchain-trusty-6.0
+            - ubuntu-toolchain-r-test
+          packages:
+            - clang-3.9
+      env:
+         - ANALYZE=false
+         - CC=clang-3.9
+         - CXX=clang++-3.9
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - llvm-toolchain-trusty-6.0
+            - ubuntu-toolchain-r-test
+          packages:
+            - clang-6.0
+            - clang-tidy-6.0
+      env:
+         - ANALYZE=false
+         - CC=clang-6.0
+         - CXX=clang++-6.0
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - gcc-7
+            - gcc-7-multilib
+            - g++-7
+            - g++-7-multilib
+            - linux-libc-dev:i386
+      env:
+         - ANALYZE=true
+         - CC=gcc-7
+         - CXX=g++-7
+    - os: linux
+      addons:
+        apt:
+          sources:
+            - llvm-toolchain-trusty-6.0
+            - ubuntu-toolchain-r-test
+          packages:
+            - clang-6.0
+            - clang-tidy-6.0
+      env:
+         - ANALYZE=true
+         - CC=clang-6.0
+         - CXX=clang++-6.0
     - os: osx
-      compiler: gcc
-    - os: osx
-      env: ANALYZE=true
+      compiler: clang
+      env:
+         - ANALYZE=false
 
-addons:
-  apt:
-    sources:
-      # see https://github.com/travis-ci/apt-source-whitelist/blob/master/ubuntu.json
-    packages:
-      - binutils-mingw-w64-i686
-      - build-essential
-      - cmake
-      - gcc-mingw-w64-i686
-      - gcc-multilib
-      - g++-multilib
-      - gcc-4.8
-      - g++-4.8
-      - g++-4.8-multilib
-      - libc6-dbg # for valgrind compilation
-      - mingw-w64
-      - g++-mingw-w64
+#addons:
+#  apt:
+#    sources:
+#      # see https://github.com/travis-ci/apt-source-whitelist/blob/master/ubuntu.json
+#    packages:
+#      - binutils-mingw-w64-i686
+#      - build-essential
+#      - cmake
+#      - mingw32-gcc
+#      - gcc-mingw-w64-i686
+#      - gcc-mingw-w64
+#      - g++-mingw-w64
+#      - gcc-multilib
+#      - g++-multilib
+#      - libc6-dbg # for valgrind compilation
 
 cache:
   pip: true

--- a/libmdnsd/1035.c
+++ b/libmdnsd/1035.c
@@ -267,7 +267,7 @@ static int _rrparse(struct message *m, struct resource *rr, int count, unsigned 
 				return 1;
 			rr[i].known.a.name = (char *)m->_packet + m->_len;
 			m->_len += 16;
-			msnds_snprintf(rr[i].known.a.name,15, "%d.%d.%d.%d", (*bufp)[0], (*bufp)[1], (*bufp)[2], (*bufp)[3]);
+			msnds_snprintf(rr[i].known.a.name,16, "%d.%d.%d.%d", (*bufp)[0], (*bufp)[1], (*bufp)[2], (*bufp)[3]);
 			rr[i].known.a.ip.s_addr = (in_addr_t)(*(*bufp) | (*(*bufp + 1) << 8) | (*(*bufp + 2) << 16) | (*(*bufp + 3) << 24));
 			break;
 

--- a/libmdnsd/1035.c
+++ b/libmdnsd/1035.c
@@ -16,6 +16,18 @@ __inline int msnds_vsnprintf(char *outBuf, size_t size, const char *format, va_l
     return count;
 }
 
+__inline int msnds_snprintf(char *outBuf, size_t size, const char *format, ...)
+{
+    int count;
+    va_list ap;
+
+    va_start(ap, format);
+    count = msnds_vsnprintf(outBuf, size, format, ap);
+    va_end(ap);
+
+    return count;
+}
+
 #else
 
 #define msnds_snprintf snprintf

--- a/libmdnsd/1035.c
+++ b/libmdnsd/1035.c
@@ -205,7 +205,8 @@ static int _host(struct message *m, unsigned char **bufp, char *name)
 			if (_lmatch(m, label + x, m->_labels[y])) {
 				/* Matching label, set up pointer */
 				l = label + x;
-				short2net((unsigned short)((unsigned char *)m->_labels[y] - m->_packet), (unsigned char **)&l);
+				unsigned char* target = (unsigned char *) &(l);
+				short2net((unsigned short)((unsigned char *)m->_labels[y] - m->_packet), &target);
 				label[x] = (char)(label[x] | 0xc0);
 				len = x + 2;
 				break;

--- a/libmdnsd/CMakeLists.txt
+++ b/libmdnsd/CMakeLists.txt
@@ -10,7 +10,7 @@ mark_as_advanced(MDNSD_COMPILE_AS_CXX)
 
 if(NOT MDNSD_COMPILE_AS_CXX AND (CMAKE_COMPILER_IS_GNUCC OR "x${CMAKE_C_COMPILER_ID}" STREQUAL "xClang"))
   add_definitions(-std=c99 -pipe
-                  -Wall -Wextra -Werror -Wpedantic
+                  -Wall -Wextra -Werror
                   -Wno-unused-parameter # some methods may require unused arguments to cast to a method pointer
                   -Wmissing-prototypes -Wstrict-prototypes -Wredundant-decls
                   -Wformat -Wformat-security -Wformat-nonliteral
@@ -21,6 +21,13 @@ if(NOT MDNSD_COMPILE_AS_CXX AND (CMAKE_COMPILER_IS_GNUCC OR "x${CMAKE_C_COMPILER
                   -Wmultichar
                   -Wundef
                   -Wc++-compat)
+  if(NOT MINGW AND NOT MSYS AND NOT MSVC)
+    add_definitions(-Wpedantic)
+  endif()
+  if(UNIX AND NOT APPLE AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 7.0)
+    # new feature since 7.0
+    add_definitions(-Wformat-truncation=1)
+  endif()
   if(NOT WIN32 AND NOT CYGWIN)
     add_definitions(-Wshadow -Wconversion -fvisibility=hidden -fPIC)
   endif()

--- a/mdnsd.c
+++ b/mdnsd.c
@@ -141,7 +141,7 @@ int msock(void)
 	int s, flag = 1, ittl = 255;
 	struct sockaddr_in in;
 	struct ip_mreq mc;
-	char ttl = 255; // send to any reachable net, not only the subnet
+	unsigned char ttl = 255; // send to any reachable net, not only the subnet
 
 	memset(&in, 0, sizeof(in));
 	in.sin_family = AF_INET;

--- a/mdnsd.c
+++ b/mdnsd.c
@@ -120,7 +120,9 @@ void done(int sig)
 	_shutdown = 1;
 	mdnsd_shutdown(_d);
 	// wake up select
-	write(daemon_socket, "\0", 1);
+	if (write(daemon_socket, "\0", 1) == -1) {
+		printf("Could not write zero byte to socket\n");
+	}
 }
 
 static void socket_set_nonblocking(int sockfd) {

--- a/tools/travis/travis_linux_before_install.sh
+++ b/tools/travis/travis_linux_before_install.sh
@@ -3,12 +3,12 @@ set -ev
 
 echo "=== Updating the build environment in $LOCAL_PKG ==="
 
-echo "=== Installing from external package sources ==="
-wget -O - http://apt.llvm.org/llvm-snapshot.gpg.key|sudo apt-key add -
-echo "deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-3.9 main" | sudo tee -a /etc/apt/sources.list
-sudo add-apt-repository -y ppa:lttng/ppa
-sudo apt-get update -qq
-sudo apt-get install -y clang-3.9 clang-tidy-3.9
+#echo "=== Installing from external package sources ==="
+#wget -O - http://apt.llvm.org/llvm-snapshot.gpg.key|sudo apt-key add -
+#echo "deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-3.9 main" | sudo tee -a /etc/apt/sources.list
+#sudo add-apt-repository -y ppa:lttng/ppa
+#sudo apt-get update -qq
+#sudo apt-get install -y clang-3.9 clang-tidy-3.9
 
 echo "=== Installing python packages ==="
 pip install --user cpp-coveralls
@@ -43,7 +43,7 @@ echo "=== Installing valgrind ==="
 wget ftp://sourceware.org/pub/valgrind/valgrind-3.13.0.tar.bz2
 tar xf valgrind-3.13.0.tar.bz2
 cd valgrind-3.13.0
-./configure --prefix=$LOCAL_PKG
+CC=clang CXX=clang++ ./configure --prefix=$LOCAL_PKG
 make -s -j8 install
 cd ..
 
@@ -66,8 +66,8 @@ echo "=== Installing cppcheck ==="
 wget https://github.com/danmar/cppcheck/archive/1.73.tar.gz -O cppcheck-1.73.tar.gz
 tar xf cppcheck-1.73.tar.gz
 cd cppcheck-1.73
-make PREFIX="$LOCAL_PKG" SRCDIR=build CFGDIR="$LOCAL_PKG/cppcheck-cfg" HAVE_RULES=yes CXXFLAGS="-O2 -DNDEBUG -Wall -Wno-sign-compare -Wno-unused-function" -j8
-make PREFIX="$LOCAL_PKG" SRCDIR=build CFGDIR="$LOCAL_PKG/cppcheck-cfg" HAVE_RULES=yes install
+CC=clang CXX=clang++ make PREFIX="$LOCAL_PKG" SRCDIR=build CFGDIR="$LOCAL_PKG/cppcheck-cfg" HAVE_RULES=yes CXXFLAGS="-O2 -DNDEBUG -Wno-sign-compare -Wno-unused-function -std=gnu++0x" -j
+CC=clang CXX=clang++ make PREFIX="$LOCAL_PKG" SRCDIR=build CFGDIR="$LOCAL_PKG/cppcheck-cfg" HAVE_RULES=yes install
 cd ..
 
 # create cached flag

--- a/tools/travis/travis_linux_script.sh
+++ b/tools/travis/travis_linux_script.sh
@@ -1,13 +1,15 @@
 #!/bin/bash
 set -ev
 
+echo "\n\n---------------------------------------------------\n###### Running with $CC and Analyze=$ANALYZE ######\n---------------------------------------------------\n\n"
+
 if [ $ANALYZE = "true" ]; then
     echo "=== Running static code analysis ==="
-    if [ "$CC" = "clang" ]; then
+    if ! case $CC in clang*) false;; esac; then
         mkdir -p build
         cd build
-        scan-build-3.9 cmake -G "Unix Makefiles" ..
-        scan-build-3.9 -enable-checker security.FloatLoopCounter \
+        scan-build cmake -G "Unix Makefiles" ..
+        scan-build -enable-checker security.FloatLoopCounter \
           -enable-checker security.insecureAPI.UncheckedReturn \
           --status-bugs -v \
           make -j
@@ -27,8 +29,7 @@ if [ $ANALYZE = "true" ]; then
 else
     echo "=== Building ==="
 
-    # cross compilation only with gcc
-    if [ "$CC" = "gcc" ]; then
+    if [ $MINGW = "true" ]; then
         echo "Cross compile release build for MinGW 32 bit"
         mkdir -p build && cd build
         cmake -DCMAKE_TOOLCHAIN_FILE=../tools/cmake/Toolchain-mingw32.cmake -DCMAKE_BUILD_TYPE=Release ..
@@ -46,11 +47,6 @@ else
         cmake -DCMAKE_TOOLCHAIN_FILE=../tools/cmake/Toolchain-gcc-m32.cmake -DCMAKE_BUILD_TYPE=Release ..
         make -j
         cd .. && rm build -rf
-    fi
-
-    if [ "$CC" = "gcc" ]; then
-        echo "Upgrade to gcc 4.8"
-        export CXX="g++-4.8" CC="gcc-4.8"
     fi
 
     echo "Compile release build for 64-bit linux"

